kafka-client/src/main/java/dev/responsive/kafka/store/ResponsiveGlobalStore.java

```diff
@@ -0,0 +1,180 @@
+ * limitations under the License.
+ */
+
+package dev.responsive.kafka.store;
+
+import dev.responsive.db.CassandraClient;
+import dev.responsive.utils.RemoteMonitor;
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.TimeoutException;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.errors.ProcessorStateException;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.StateStoreContext;
+import org.apache.kafka.streams.query.Position;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.internals.StoreQueryUtils;
+
+/**
+ * The {@code ResponsiveGlobalStore}
+ */
+public class ResponsiveGlobalStore implements KeyValueStore<Bytes, byte[]> {
+
+  private final CassandraClient client;
+  private final String name;
+  private final String tableName;
+  private final RemoteMonitor initRemote;
+  private final Position position;
+
+  private boolean open;
+  private int partition;
+  private StateStoreContext context;
+
+  public ResponsiveGlobalStore(
+      final CassandraClient client,
+      final String name,
+      final RemoteMonitor initRemote
+  ) {
+    this.client = client;
+    this.name = name;
+    this.tableName = '"' + name + '"';
+    this.initRemote = initRemote;
+    this.position = Position.emptyPosition();
+  }
+
+  @Override
+  public String name() {
+    return name;
+  }
+
+  @Override
+  @SuppressWarnings("deprecation")
+  public void init(final ProcessorContext context, final StateStore root) {
+    if (context instanceof StateStoreContext) {
+      init((StateStoreContext) context, root);
+    } else {
+      throw new UnsupportedOperationException(
+          "Use ResponsiveStore#init(StateStoreContext, StateStore) instead."
+      );
+    }
+  }
+
+  @Override
+  public void init(final StateStoreContext context, final StateStore root) {
+    try {
+      this.context = context;
+      // this is bad, but the assumption is that global tables are small
+      // and can fit in a single partition - all writers will write using
+      // the same partition key. we should consider using ALLOW FILTERING
+      // instead. that would make gets/puts more efficient and the queries
+      // to be more evently distributed and take the hit only on range
+      // queries
+      partition = 0;
+      client.createDataTable(tableName);
+      initRemote.await(Duration.ofSeconds(60));
+
+      client.prepareStatements(tableName);
+      client.initializeOffset(tableName, partition);
+
+      open = true;
+
+      context.register(root, (key, value) -> {
+        if (value == null) {
+          delete(new Bytes(key));
+        } else {
+          put(new Bytes(key), value);
+        }
```

Comment:
Should we be worried about the overhead of all these single non-batched writes to C*? I know global stores are small, updated infrequently, etc...but loading data from an existing global table will make the migration/restoration period rough 😬

Can we implement a batching version of `putAll()`? If so, we can pass in a `RecordBatchingStateRestoreCallback` here that restores batches with all the records returned by `poll`. 

You probably noticed the GlobalStateManagerImpl invokes `restoreBatch` but that gets split into individual records depending on which version of the `StateRestoreCallback` is passed in right here (it's kind of a ridiculous API)

Reply:
Cassandra/Scylla actually do better without batched writes (they are for atomicity, not performance): https://docs.datastax.com/en/cql-oss/3.x/cql/cql_using/useBatch.html so this should be better

Reply:
FWIW I found this blog series to provide a good (and short!) high-level summary: [post #1](https://foundev.medium.com/cassandra-batch-loading-without-the-batch-keyword-40f00e35e23e) and [post #2](https://foundev.medium.com/cassandra-batch-loading-without-the-batch-the-nuanced-edition-dd78d61e9885)

The 2nd one has some solid memes

Reply:
Heh, I had spent quite a bit of time reading up on batching etc.. in scylla/cassandra when researching it for the CAS use case we had so I'm aware of logged/unlogged batching and the limitations around single-partitions/cross-partition writes.

My opinion here is that we shouldn't use batches in hope of improved performance because that's not what they're meant for. There might be situations (as described in post 2) where they will improve the performance, but I would argue that ideally, this kind of performance batching is a client optimization that is provided by the vendor. I see no reason why the Scylla driver can't eventually come up with a "write buffer size" configuration that batches writes to the network, and if we explicitly used Cassandra Batches we would not be able to take advantage of such a feature.

I do agree with your assessment that we should keep an eye out for this, and if global state restoration proves to be an issue we should work with our contacts at Scylla to figure out optimization - but for now I think changing semantics would be premature optimization.

Reply:
Also I think we should eventually move away from global partitions being a single partition in the remote store - that just made this PR a bit simpler because I could reuse all the existing client code for things like range scans, but it's not a requirement (we could create an iterator across each partition when doing global lookups, or use `ALLOW_FILTERING`).

Reply:
https://opensource.docs.scylladb.com/stable/operating-scylla/procedures/tips/benchmark-tips.html#query-recommendations hmm, maybe I didn't understand this properly, as they do actually recommend batching for performance in this document